fcreatedb/fcreatedb.1
.\" El manpage-o for fcreatedb
.\" (C)2005, Dan Ponte
.\" $Amigan: fakedbfs/fcreatedb/fcreatedb.1,v 1.2 2005/09/24 17:03:07 dcp1990 Exp $
.Dd August 13, 2005
.Dt FCREATEDB 1
.Os
.Sh NAME
.Nm fcreatedb
.Nd initialise/append to fakedbfs databases from specfiles
.Sh SYNOPSIS
.Nm
.Op Fl hva
.Op Fl d Ar dbfile
.Ar specfile
.Sh DESCRIPTION
.Nm
is part of the
.Em fakedbfs
distribution. It is used to prepare databases for use by the various other
utilities in the distribution.
.Pp
.Nm
accepts one mandatory argument,
.Ar specfile .
This file's format is that of
.Xr specfile 5 .
It specifies the various catalogue fields and enum declarations of the database.
.Pp
One can also use the
.Ar d
argument to specify the database to use; by default it is either the value of
the FAKEDBFSDB environment variable, or $HOME/.fakedbfsdb. This convention is
followed by all utilities in this distribution, and should be followed for other
applications using the libfakedbfs library.
.Pp
.Nm
recognizes the following command line options:
.Bl -tag -width indent
.It Fl h
Print a short usage message.
.It Fl a
Put
.Nm
in
.Dq append
mode. This means that
.Nm
won't even attempt to create the database and set up the default tables (which
is the default). This option is quite useless, as it already checks if the
database is initialised before attempting anyway. It is provided primarily for
debugging purposes.
.It Fl v
Print the library version, the
.Nm
utility version, and other information.
.It Fl d Ar dbfile
Use
.Ar dbfile
as the database instead of the default.
.El
.Sh DIAGNOSTICS
The
.Nm
utility will print whatever error messages it gets from the library to
.Em stderr .
These usually will include a recursing list of errors as each function returns;
see
.Xr libfakedbfs 3
for more information.
.Sh ENVIRONMENT
The FAKEDBFSDB is the only environment variable that
.Nm
uses. See above for use information.
.Sh FILES
.Bl -tag -width fdbfs
.It Pa ~/.fakedbfsdb
default database file
.El
.Sh SEE ALSO
.Xr libfakedbfs 3 ,
.Xr specfile 5 ,
.Xr findex 1 ,
.Xr fedit 1 ,
.Xr fquery 1
.Sh HISTORY
The
.Nm
utility appeared in fakedbfs v1.0.
.Sh BUGS
.Nm
should be merged into
.Xr fedit 1 ,
but the author is too lazy.
.Sh AUTHOR
Dan Ponte <dcp1990@neptune.atopia.net>
.Pp
Web: http://www.theamigan.net/fakedbfs/


fcreatedb/fcreatedb.c
/*
 * Copyright (c) 2005, Dan Ponte
 *
 * fcreatedb.c - fcreate main (for creating databases)
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the author nor the names of his contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
/* $Amigan: fakedbfs/fcreatedb/fcreatedb.c,v 1.11 2005/09/06 07:27:22 dcp1990 Exp $ */
/* system includes */
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#include <fakedbfs.h>
#include <fakedbfsapps.h>

#define ARGSPEC "vhad:"
#define FCREATEVER "0.1"

RCSID("$Amigan: fakedbfs/fcreatedb/fcreatedb.c,v 1.11 2005/09/06 07:27:22 dcp1990 Exp $")

static int dbfu = 0;
static int append = 0;
char *dbf, *specf;
fdbfs_t *f;

void version(void)
{
	printf("fcreatedb v%s. (C)2005, Dan Ponte.\n"
		       "Under the BSD license; see the source for more details.\n"
		       "fakedbfs library v%s ('%s'). Originally built for v%s.\n%s\n"
		       "Visit http://www.theamigan.net/fakedbfs/ for more info.\n", FCREATEVER, fakedbfsver,
		      fakedbfsvname, FAKEDBFSVER, fakedbfscopyright);
}
void usage(pn)
	char *pn;
{
	fprintf(stderr, "%s: usage: %s [-d dbfile] [-a] [-h] [-v] specfile\n",
			pn, pn);
}

int main(argc, argv)
	int argc;
	char *argv[];
{
	int c;
	char *estr;
	char *pname;
	
	pname = strdup(argv[0]);

	while((c = getopt(argc, argv, ARGSPEC)) != -1)
		switch(c) {
			case 'v':
				version();
				free(pname);
				return 0;
			case 'h':
				usage(pname);
				free(pname);
				return 0;
			case 'd':
				dbf = strdup(optarg);
				dbfu = 1;
				break;
			case 'a':
				append = 1;
				break;
			case '?':
			default:
				usage(pname);
				free(pname);
				return 1;
		}
	argc -= optind;
	argv += optind;

	if(argc != 1) {
		usage(pname);
		free(pname);
		return -1;
	}
	
	specf = strdup(argv[0]);
	
	free(pname);
	
	if(!dbfu) {
		if(getenv(FDBFSDBENV) != NULL)
			dbf = strdup(getenv(FDBFSDBENV));
		else
			asprintf(&dbf, "%s/.fakedbfsdb", getenv("HOME"));
	}
	
	f = new_fdbfs(dbf, &estr, DEBUGFUNC_STDERR, 0);
	if(f == NULL) {
		fprintf(stderr, "error creating fdbfs instance: %s\n", estr);
		free(dbf);
		free(estr);
		free(specf);
		return -1;
	}
	if(!append)
		if(!start_db(f)) {
			fprintf(stderr, "starting DB: %s\n", f->error.emsg);
			estr_free(&f->error);
			free(dbf);
			destroy_fdbfs(f);
			free(specf);
			return -1;
		}

	printf("Reading specfile %s...\n", specf);

	if(!parse_definition(f, specf)) {
		fprintf(stderr, "Error parsing: %s\n", f->error.emsg);
		estr_free(&f->error);
		destroy_fdbfs(f);
		free(dbf);
		free(specf);
		return -1;
	}

	printf("Database updated.\n");

	destroy_fdbfs(f);
	free(dbf);
	free(specf);

	return 0;
}
